File: resources/ios/Sources/AnnotatorCanvasView.swift
import UIKit

protocol AnnotatorCanvasViewHost: AnyObject {
    /// True while the inline text editor is open.
    func canvasIsEditingText() -> Bool
    /// A touch landed outside the text editor: commit it.
    func canvasCommitTextEditing()
    func canvasBeginTextEditing(_ session: Pteal79Annotator.TextEditSession)
    func canvasEditorChanged()
}

/// Shows the image with its annotations and turns touches into editor calls.
/// The image sits in an image view; the shapes view draws the annotations in
/// draw(_:) through one image-to-view transform.
///
/// One finger draws, selects or edits. Two fingers pinch-zoom and pan; the
/// second finger drops any gesture in progress and never starts a shape. In
/// Select mode, one finger on empty space pans while zoomed in.
final class AnnotatorCanvasView: UIView {
    typealias Editor = Pteal79Annotator.Editor

    static let maxZoom: CGFloat = 8
    static let zoomStep: CGFloat = 1.5

    private enum Mode { case none, draw, pan, zoom, swallow }

    weak var host: AnnotatorCanvasViewHost?

    private let editor: Editor
    private let imageView = UIImageView()
    private let shapesView: AnnotatorShapesView
    private let placeholder = UILabel()

    var image: UIImage? {
        didSet {
            imageView.image = image
            placeholder.isHidden = image != nil
            zoom = 1
            imageRect = .zero
            setNeedsLayout()
            layoutIfNeeded()
            shapesView.setNeedsDisplay()
        }
    }

    /// Where the image sits in this view, in points.
    private(set) var imageRect: CGRect = .zero

    /// 1 when the image is fitted, up to maxZoom.
    private(set) var zoom: CGFloat = 1

    var canZoomIn: Bool { image != nil && zoom < Self.maxZoom - 0.001 }
    var canZoomOut: Bool { image != nil && zoom > 1.001 }

    /// View points per image pixel with the image fitted.
    private var fitViewPerImage: CGFloat = 1
    private var viewPerImage: CGFloat { fitViewPerImage * zoom }
    private var lastSize: CGSize = .zero

    private var mode = Mode.none
    /// Touches in the order they came down, so the pinch pair stays stable.
    private var touches: [UITouch] = []
    private var lastPoint: CGPoint = .zero
    private var lastSpan: CGFloat = 0

    init(editor: Editor) {
        self.editor = editor
        self.shapesView = AnnotatorShapesView(editor: editor)
        super.init(frame: .zero)

        backgroundColor = UIColor(red: 15 / 255, green: 15 / 255, blue: 15 / 255, alpha: 1)
        isMultipleTouchEnabled = true
        clipsToBounds = true

        imageView.contentMode = .scaleToFill
        addSubview(imageView)

        shapesView.backgroundColor = .clear
        shapesView.isOpaque = false
        shapesView.isUserInteractionEnabled = false
        shapesView.contentMode = .redraw
        addSubview(shapesView)

        placeholder.text = "Loading image\u{2026}"
        placeholder.textColor = UIColor(white: 0.62, alpha: 1)
        placeholder.font = .systemFont(ofSize: 15)
        placeholder.textAlignment = .center
        addSubview(placeholder)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    /// Fits the image, applies zoom and pan, and recomputes the editor's scales. Shapes never change.
    override func layoutSubviews() {
        super.layoutSubviews()
        placeholder.frame = bounds
        shapesView.frame = bounds

        guard let cgImage = image?.cgImage, bounds.width > 0, bounds.height > 0 else { return }
        let pixelWidth = CGFloat(cgImage.width)
        let pixelHeight = CGFloat(cgImage.height)

        // Keep the image point at the centre of the old size in the centre.
        var centre: CGPoint?
        if imageRect != .zero, lastSize != bounds.size, lastSize.width > 0 {
            centre = toImage(CGPoint(x: lastSize.width / 2, y: lastSize.height / 2))
        }
        let resized = lastSize != bounds.size
        lastSize = bounds.size

        fitViewPerImage = min(bounds.width / pixelWidth, bounds.height / pixelHeight)
        let size = CGSize(width: pixelWidth * viewPerImage, height: pixelHeight * viewPerImage)

        var origin = imageRect.origin
        if let centre {
            origin = CGPoint(x: bounds.width / 2 - centre.x * viewPerImage, y: bounds.height / 2 - centre.y * viewPerImage)
        } else if imageRect == .zero {
            origin = CGPoint(x: (bounds.width - size.width) / 2, y: (bounds.height - size.height) / 2)
        }
        let before = imageRect
        imageRect = clamped(CGRect(origin: origin, size: size))

        // Image pixels per screen point, fitted and at the current zoom.
        editor.scale = pixelWidth / (pixelWidth * fitViewPerImage)
        editor.touchScale = editor.scale / zoom
        applyImageRect()
        if resized || imageRect != before { host?.canvasEditorChanged() }
    }

    /// Centres the image on an axis where it is smaller than the view, otherwise keeps the view covered.
    private func clamped(_ rect: CGRect) -> CGRect {
        var r = rect
        r.origin.x = r.width <= bounds.width ? (bounds.width - r.width) / 2 : min(max(r.minX, bounds.width - r.width), 0)
        r.origin.y = r.height <= bounds.height ? (bounds.height - r.height) / 2 : min(max(r.minY, bounds.height - r.height), 0)
        return r
    }

    private func applyImageRect() {
        imageView.frame = imageRect
        shapesView.imageRect = imageRect
        shapesView.setNeedsDisplay()
    }

    /// Zooms by factor, keeping the image point under focus still.
    func zoom(by factor: CGFloat, focus: CGPoint? = nil) {
        guard let cgImage = image?.cgImage else { return }
        let next = min(max(zoom * factor, 1), Self.maxZoom)
        if next == zoom { return }

        let f = focus ?? CGPoint(x: bounds.midX, y: bounds.midY)
        let anchor = toImage(f)
        zoom = next
        let size = CGSize(width: CGFloat(cgImage.width) * viewPerImage, height: CGFloat(cgImage.height) * viewPerImage)
        imageRect = clamped(CGRect(x: f.x - anchor.x * viewPerImage, y: f.y - anchor.y * viewPerImage, width: size.width, height: size.height))
        editor.touchScale = editor.scale / zoom
        applyImageRect()
        host?.canvasEditorChanged()
    }

    func resetZoom() {
        guard image != nil, zoom != 1 else { return }
        zoom = 1
        imageRect = .zero
        setNeedsLayout()
        layoutIfNeeded()
        host?.canvasEditorChanged()
    }

    private func pan(by dx: CGFloat, _ dy: CGFloat) {
        imageRect = clamped(imageRect.offsetBy(dx: dx, dy: dy))
        applyImageRect()
        host?.canvasEditorChanged()
    }

    func toImage(_ p: CGPoint) -> CGPoint {
        CGPoint(x: (p.x - imageRect.minX) / viewPerImage, y: (p.y - imageRect.minY) / viewPerImage)
    }

    func toView(_ p: CGPoint) -> CGPoint {
        CGPoint(x: imageRect.minX + p.x * viewPerImage, y: imageRect.minY + p.y * viewPerImage)
    }

    func redraw() {
        shapesView.setNeedsDisplay()
    }

    // MARK: Touches

    override func touchesBegan(_ began: Set<UITouch>, with event: UIEvent?) {
        guard image != nil, isUserInteractionEnabled else { return }
        let wasEmpty = touches.isEmpty
        touches.append(contentsOf: began)

        if wasEmpty, let touch = touches.first {
            lastPoint = touch.location(in: self)
            if host?.canvasIsEditingText() == true {
                host?.canvasCommitTextEditing()
                mode = .swallow
            } else {
                editor.touchDown(toImage(lastPoint))
                mode = editor.touchIsIdle && zoom > 1 ? .pan : .draw
                changed()
            }
        }

        if touches.count >= 2 && mode != .zoom {
            // A second finger never draws: drop whatever the first one started.
            if mode == .draw { editor.abortTouch() }
            mode = .zoom
            startPinch()
            changed()
        }
    }

    override func touchesMoved(_ moved: Set<UITouch>, with event: UIEvent?) {
        guard image != nil, let first = touches.first else { return }

        switch mode {
        case .draw:
            let samples = event?.coalescedTouches(for: first) ?? [first]
            for sample in samples {
                editor.touchMove(toImage(sample.location(in: self)))
            }
            changed()
        case .pan:
            let p = first.location(in: self)
            pan(by: p.x - lastPoint.x, p.y - lastPoint.y)
            lastPoint = p
        case .zoom:
            guard touches.count >= 2 else { return }
            let (focus, span) = pinch()
            pan(by: focus.x - lastPoint.x, focus.y - lastPoint.y)
            if lastSpan > 0, span > 0 { zoom(by: span / lastSpan, focus: focus) }
            lastSpan = span
            lastPoint = focus
        case .none, .swallow:
            break
        }
    }

    override func touchesEnded(_ ended: Set<UITouch>, with event: UIEvent?) {
        guard image != nil else { return }
        let first = touches.first
        touches.removeAll { ended.contains($0) }

        if touches.isEmpty {
            if (mode == .draw || mode == .pan), let first {
                let session = editor.touchUp(toImage(first.location(in: self)))
                changed()
                if let session { host?.canvasBeginTextEditing(session) }
            }
            mode = .none
        } else if mode == .zoom {
            // Carry on pinching with the fingers that are left, or wait for them to lift.
            if touches.count >= 2 { startPinch() } else { mode = .swallow }
        }
    }

    override func touchesCancelled(_ cancelled: Set<UITouch>, with event: UIEvent?) {
        guard image != nil else { return }
        touches.removeAll { cancelled.contains($0) }
        guard touches.isEmpty else { return }

        if mode == .draw { editor.touchCancel() } else if mode == .pan { editor.abortTouch() }
        mode = .none
        changed()
    }

    private func startPinch() {
        let (focus, span) = pinch()
        lastPoint = focus
        lastSpan = span
    }

    private func pinch() -> (focus: CGPoint, span: CGFloat) {
        let a = touches[0].location(in: self)
        let b = touches[1].location(in: self)
        return (CGPoint(x: (a.x + b.x) / 2, y: (a.y + b.y) / 2), hypot(a.x - b.x, a.y - b.y))
    }

    private func changed() {
        shapesView.setNeedsDisplay()
        host?.canvasEditorChanged()
    }
}

/// Draws the annotations over the image view, in image pixels.
private final class AnnotatorShapesView: UIView {
    private let editor: Pteal79Annotator.Editor
    var imageRect: CGRect = .zero

    init(editor: Pteal79Annotator.Editor) {
        self.editor = editor
        super.init(frame: .zero)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func draw(_ rect: CGRect) {
        guard let ctx = UIGraphicsGetCurrentContext(), imageRect.width > 0 else { return }

        // touchScale is image pixels per point at the current zoom.
        ctx.saveGState()
        ctx.translateBy(x: imageRect.minX, y: imageRect.minY)
        ctx.scaleBy(x: 1 / editor.touchScale, y: 1 / editor.touchScale)

        let shapes = editor.shapes
        let preview = editor.preview
        let hidden = editor.hiddenIndex
        for (i, shape) in shapes.enumerated() where i != hidden {
            let drawn = preview?.index == i ? preview!.shape : shape
            Pteal79Annotator.Renderer.draw(drawn, in: ctx)
        }

        if let draft = editor.draft {
            Pteal79Annotator.Renderer.draw(draft, in: ctx)
        }

        if let i = editor.selectedIndex, i != hidden, shapes.indices.contains(i) {
            let shape = preview?.index == i ? preview!.shape : shapes[i]
            Pteal79Annotator.Renderer.drawOverlay(shape, scale: editor.touchScale, in: ctx)
        }

        ctx.restoreGState()
    }
}

File: resources/android/src/AnnotatorCanvasView.kt
package com.pteal79.plugins.imageannotator

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.PointF
import android.graphics.RectF
import android.view.MotionEvent
import android.view.View
import kotlin.math.hypot
import kotlin.math.min

/**
 * Shows the image with its annotations and turns touches into editor calls.
 * All drawing happens in image pixels through one transform.
 *
 * One finger draws, selects or edits. Two fingers pinch-zoom and pan; the
 * second finger drops any gesture in progress and never starts a shape. In
 * Select mode, one finger on empty space pans while zoomed in.
 */
@SuppressLint("ViewConstructor")
class AnnotatorCanvasView(
    context: Context,
    private val editor: AnnotatorEditor,
    private val renderer: AnnotatorRenderer,
) : View(context) {

    interface Host {
        /** True while the inline text editor is open. */
        fun isEditingText(): Boolean

        /** A touch landed outside the text editor: commit it. */
        fun commitTextEditing()

        fun beginTextEditing(session: TextEditSession)

        fun onEditorChanged()
    }

    private enum class Mode { NONE, DRAW, PAN, ZOOM, SWALLOW }

    var host: Host? = null

    var image: Bitmap? = null
        set(value) {
            field = value
            zoom = 1f
            imageRect.setEmpty()
            updateLayout()
            invalidate()
        }

    /** Where the image sits in this view, in view pixels. */
    val imageRect = RectF()

    /** 1 when the image is fitted, up to MAX_ZOOM. */
    var zoom = 1f
        private set

    val canZoomIn: Boolean get() = image != null && zoom < MAX_ZOOM - 0.001f
    val canZoomOut: Boolean get() = image != null && zoom > 1.001f

    /** View pixels per image pixel with the image fitted. */
    private var fitViewPerImage = 1f

    /** View pixels per image pixel at the current zoom. */
    private val viewPerImage: Float get() = fitViewPerImage * zoom

    private val density = resources.displayMetrics.density
    private val imagePaint = Paint(Paint.FILTER_BITMAP_FLAG or Paint.ANTI_ALIAS_FLAG)
    private val placeholderPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = Color.parseColor("#9ca3af")
        textSize = 15f * density
        textAlign = Paint.Align.CENTER
    }

    private var mode = Mode.NONE
    private var lastX = 0f
    private var lastY = 0f
    private var lastSpan = 0f

    init {
        setBackgroundColor(Color.parseColor("#0f0f0f"))
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        // Keep the image point at the centre of the old size in the centre.
        val centre = if (!imageRect.isEmpty && oldw > 0 && oldh > 0) toImage(oldw / 2f, oldh / 2f) else null
        updateLayout(centre)
    }

    /** Fits the image, applies zoom and pan, and recomputes the editor's scales. Shapes never change. */
    private fun updateLayout(keepCentred: Pt? = null) {
        val bitmap = image ?: return
        if (width == 0 || height == 0) return

        fitViewPerImage = min(width.toFloat() / bitmap.width, height.toFloat() / bitmap.height)
        val w = bitmap.width * viewPerImage
        val h = bitmap.height * viewPerImage

        when {
            keepCentred != null -> imageRect.set(
                width / 2f - keepCentred.x * viewPerImage,
                height / 2f - keepCentred.y * viewPerImage,
                0f,
                0f,
            )
            imageRect.isEmpty -> imageRect.set((width - w) / 2f, (height - h) / 2f, 0f, 0f)
        }
        imageRect.right = imageRect.left + w
        imageRect.bottom = imageRect.top + h
        clampPan()

        // Image pixels per screen point (dp), fitted and at the current zoom.
        editor.scale = bitmap.width / (bitmap.width * fitViewPerImage / density)
        editor.touchScale = editor.scale / zoom
        host?.onEditorChanged()
    }

    /** Centres the image on an axis where it is smaller than the view, otherwise keeps the view covered. */
    private fun clampPan() {
        val w = imageRect.width()
        val h = imageRect.height()
        val left = if (w <= width) (width - w) / 2f else imageRect.left.coerceIn(width - w, 0f)
        val top = if (h <= height) (height - h) / 2f else imageRect.top.coerceIn(height - h, 0f)
        imageRect.set(left, top, left + w, top + h)
    }

    /** Zooms by factor, keeping the image point under (focusX, focusY) still. */
    fun zoomBy(factor: Float, focusX: Float = width / 2f, focusY: Float = height / 2f) {
        val bitmap = image ?: return
        val next = (zoom * factor).coerceIn(1f, MAX_ZOOM)
        if (next == zoom) return

        val anchor = toImage(focusX, focusY)
        zoom = next
        val left = focusX - anchor.x * viewPerImage
        val top = focusY - anchor.y * viewPerImage
        imageRect.set(left, top, left + bitmap.width * viewPerImage, top + bitmap.height * viewPerImage)
        clampPan()
        editor.touchScale = editor.scale / zoom
        changed()
    }

    fun resetZoom() {
        if (image == null || zoom == 1f) return
        zoom = 1f
        imageRect.setEmpty()
        updateLayout()
        invalidate()
    }

    private fun panBy(dx: Float, dy: Float) {
        imageRect.offset(dx, dy)
        clampPan()
        changed()
    }

    fun toImage(x: Float, y: Float) = Pt((x - imageRect.left) / viewPerImage, (y - imageRect.top) / viewPerImage)

    fun toView(p: Pt) = PointF(imageRect.left + p.x * viewPerImage, imageRect.top + p.y * viewPerImage)

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        val bitmap = image
        if (bitmap == null) {
            canvas.drawText("Loading image…", width / 2f, height / 2f, placeholderPaint)
            return
        }

        canvas.save()
        canvas.translate(imageRect.left, imageRect.top)
        canvas.scale(viewPerImage, viewPerImage)

        canvas.drawBitmap(bitmap, 0f, 0f, imagePaint)

        val shapes = editor.shapes
        val preview = editor.preview
        val hidden = editor.hiddenIndex
        shapes.forEachIndexed { i, shape ->
            if (i == hidden) return@forEachIndexed
            renderer.draw(canvas, if (preview?.first == i) preview.second else shape)
        }

        editor.draft?.let { renderer.draw(canvas, it) }

        editor.selectedIndex?.let { i ->
            if (i != hidden) {
                val shape = if (preview?.first == i) preview.second else shapes.getOrNull(i)
                shape?.let { renderer.drawOverlay(canvas, it, editor.touchScale) }
            }
        }

        canvas.restore()
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (image == null || !isEnabled) return false

        when (event.actionMasked) {
            MotionEvent.ACTION_DOWN -> {
                lastX = event.x
                lastY = event.y
                if (host?.isEditingText() == true) {
                    host?.commitTextEditing()
                    mode = Mode.SWALLOW
                    return true
                }
                editor.touchDown(toImage(event.x, event.y))
                mode = if (editor.touchIsIdle && zoom > 1f) Mode.PAN else Mode.DRAW
                changed()
            }

            MotionEvent.ACTION_POINTER_DOWN -> {
                // A second finger never draws: drop whatever the first one started.
                if (mode == Mode.DRAW) editor.abortTouch()
                mode = Mode.ZOOM
                startPinch(event)
                changed()
            }

            MotionEvent.ACTION_MOVE -> when (mode) {
                Mode.DRAW -> {
                    for (h in 0 until event.historySize) {
                        editor.touchMove(toImage(event.getHistoricalX(0, h), event.getHistoricalY(0, h)))
                    }
                    editor.touchMove(toImage(event.x, event.y))
                    changed()
                }
                Mode.PAN -> {
                    panBy(event.x - lastX, event.y - lastY)
                    lastX = event.x
                    lastY = event.y
                }
                Mode.ZOOM -> if (event.pointerCount >= 2) {
                    val span = span(event)
                    val fx = (event.getX(0) + event.getX(1)) / 2f
                    val fy = (event.getY(0) + event.getY(1)) / 2f
                    panBy(fx - lastX, fy - lastY)
                    if (lastSpan > 0f && span > 0f) zoomBy(span / lastSpan, fx, fy)
                    lastSpan = span
                    lastX = fx
                    lastY = fy
                }
                else -> {}
            }

            MotionEvent.ACTION_POINTER_UP -> {
                if (mode == Mode.ZOOM && event.pointerCount > 2) {
                    // Carry on pinching with the fingers that are left.
                    startPinch(event, skip = event.actionIndex)
                } else {
                    mode = Mode.SWALLOW
                }
            }

            MotionEvent.ACTION_UP -> {
                if (mode == Mode.DRAW || mode == Mode.PAN) {
                    val session = editor.touchUp(toImage(event.x, event.y))
                    changed()
                    session?.let { host?.beginTextEditing(it) }
                }
                mode = Mode.NONE
            }

            MotionEvent.ACTION_CANCEL -> {
                if (mode == Mode.DRAW) editor.touchCancel() else if (mode == Mode.PAN) editor.abortTouch()
                mode = Mode.NONE
                changed()
            }
        }
        return true
    }

    private fun startPinch(event: MotionEvent, skip: Int = -1) {
        val indices = (0 until event.pointerCount).filter { it != skip }.take(2)
        if (indices.size < 2) return
        val (a, b) = indices
        lastSpan = hypot(event.getX(a) - event.getX(b), event.getY(a) - event.getY(b))
        lastX = (event.getX(a) + event.getX(b)) / 2f
        lastY = (event.getY(a) + event.getY(b)) / 2f
    }

    private fun span(event: MotionEvent) = hypot(event.getX(0) - event.getX(1), event.getY(0) - event.getY(1))

    private fun changed() {
        invalidate()
        host?.onEditorChanged()
    }

    companion object {
        const val MAX_ZOOM = 8f
        const val ZOOM_STEP = 1.5f
    }
}
